sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/PartitionKeyBasedBloomFilter.java

```diff
@@ -0,0 +1,139 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.cosmos.implementation;
+
+import com.azure.cosmos.implementation.guava25.hash.BloomFilter;
+import com.azure.cosmos.implementation.guava25.hash.Funnel;
+import com.azure.cosmos.implementation.guava25.hash.PrimitiveSink;
+import com.azure.cosmos.implementation.routing.PartitionKeyInternal;
+import com.azure.cosmos.implementation.routing.PartitionKeyInternalHelper;
+import com.azure.cosmos.models.PartitionKeyDefinition;
+
+import java.nio.charset.StandardCharsets;
+import java.util.HashSet;
+import java.util.Locale;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public class PartitionKeyBasedBloomFilter {
+
+    private static final long EXPECTED_INSERTIONS = Configs.getPkBasedBloomFilterExpectedInsertionCount();
+    private static final double ALLOWED_FALSE_POSITIVE_RATE = Configs.getPkBasedBloomFilterExpectedFfpRate();
+    private BloomFilter<PartitionKeyBasedBloomFilterType> pkBasedBloomFilter;
+    private final Set<String> recordedRegions;
+    private final Funnel<PartitionKeyBasedBloomFilterType> funnel;
+
+    public PartitionKeyBasedBloomFilter() {
+        this.recordedRegions = ConcurrentHashMap.newKeySet();
+        this.funnel = new Funnel<PartitionKeyBasedBloomFilterType>() {
+            @Override
+            public void funnel(PartitionKeyBasedBloomFilterType from, PrimitiveSink into) {
+                into
+                    .putLong(from.collectionRid)
+                    .putString(from.effectivePartitionKeyString, StandardCharsets.UTF_8)
+                    .putString(from.region, StandardCharsets.UTF_8);
+            }
+        };
+        this.pkBasedBloomFilter = BloomFilter.create(this.funnel, EXPECTED_INSERTIONS, ALLOWED_FALSE_POSITIVE_RATE);
+    }
+
+    public void tryRecordPartitionKey(
+        RxDocumentServiceRequest request,
+        Long collectionRid,
+        String firstPreferredWritableRegion,
+        String regionRoutedTo,
+        PartitionKeyInternal partitionKeyInternal,
+        PartitionKeyDefinition partitionKeyDefinition) {
+
+        try {
+            if (partitionKeyInternal == null) {
+                return;
+            }
+
+            if (partitionKeyDefinition == null) {
+                return;
+            }
+
+            if (Strings.isNullOrEmpty(firstPreferredWritableRegion)) {
+                return;
+            }
+
+
+            String effectivePartitionKeyString = request.getEffectivePartitionKey() != null ? request.getEffectivePartitionKey() : PartitionKeyInternalHelper
+                .getEffectivePartitionKeyString(partitionKeyInternal, partitionKeyDefinition);
+
+            String normalizedRegionRoutedTo = regionRoutedTo.toLowerCase(Locale.ROOT).replace(" ", "");;
```

Comment:
Can we avoid calculating this every time - for example having a lookup dictionary?

Reply:
Fixed in next iteration.